src/types.cc

```diff
@@ -807,6 +810,175 @@ namespace LogReg {
     lua_setglobal(L, "log");
   }
 }
+namespace DictEntryReg {
+  typedef DictEntry T;
+  an<T> make() {
+    return an<T>(new T());
+  }
+
+  static const luaL_Reg funcs[] = {
+    {"DictEntry",WRAP(make)},
+    { NULL, NULL },
+  };
+
+  static const luaL_Reg methods[] = {
+    { NULL, NULL },
+  };
+
+  static const luaL_Reg vars_get[] = {
+    {"text",WRAPMEM_GET(T::text)},
+    {"comment",WRAPMEM_GET(T::comment)},
+    {"preedit",WRAPMEM_GET(T::preedit)},
+    {"weight",WRAPMEM_GET(T::weight)},
+    {"commit_count", WRAPMEM_GET(T::commit_count)},
+    {"custom_code",WRAPMEM_GET(T::custom_code)},
+    {"remaining_code_length",WRAPMEM_GET(T::remaining_code_length)},
+    {"code",WRAPMEM_GET(T::code)},
+    { NULL, NULL },
+  };
+
+  static const luaL_Reg vars_set[] = {
+    {"text",WRAPMEM_SET(T::text)},
+    {"comment",WRAPMEM_SET(T::comment)},
+    {"preedit",WRAPMEM_SET(T::preedit)},
+    {"weight",WRAPMEM_SET(T::weight)},
+    {"commit_count", WRAPMEM_SET(T::commit_count)},
+    {"custom_code",WRAPMEM_SET(T::custom_code)},
+    {"remaining_code_length",WRAPMEM_SET(T::remaining_code_length)},
+    {"code",WRAPMEM_SET(T::code)},
+    { NULL, NULL },
+  };
+}
+namespace CodeReg {
+
+  typedef Code T;
+  using SyllableId = int32_t; // copy from rime/vocabulary.h
+
+  an<T> make() {
+    return an<T>(new Code());
+  }
+  
+  void pushCode(T& code,const SyllableId inputCode) {
+    code.push_back(inputCode);
+  }
+
+  string printCode(T& code) {
+    return code.ToString();
+  }
+
+  static const luaL_Reg funcs[] = {
+    {"Code", WRAP(make)},
+    { NULL, NULL },
+  };
+
+  static const luaL_Reg methods[] = {
+    {"push",WRAP(pushCode)},
+    {"print",WRAP(printCode)},
+    { NULL, NULL },
+  };
+
+  static const luaL_Reg vars_get[] = {
+    { NULL, NULL },
+  };
+
+  static const luaL_Reg vars_set[] = {
+    { NULL, NULL },
+  };
+}
+namespace MemoryReg {
+  using SyllableId = int32_t; // copy from rime/vocabulary.h
+
+  class LuaMemory : public Memory {
```

Comment:
I'm not familiar with `Memory`. Can you add a example script to demo how to use it?

Reply:
here's my demo script for memory right now;
```lua
   mem = Memory(env.ticket)
   mem:dictLookup("onf",false)
   while (mem:dictResultExhausted() != false)
   do
       dict_entry = mem:dictPeek()
       print(dict_entry.text)
       print(dict_entry.weight)
       print(dict_entry.preedit)
       print(dict_entry.comment)
       ...
   end
```

Reply:
> I'm not familiar with `Memory`. Can you add a example script to demo how to use it?

I'm about to write a doc about librime-Lua, and I will add the usage of Memory and other components to the doc

Reply:
Can we wrap the dict class directly instead of the `Memory` class? In the example, the `Memory` object is only used to iterator a dict. I still want to know what `Memory` is.

Reply:
Okay. Memory class is a set of dictionary and UserDictionary for this schema,  but there's more.
Memory has a method named `memorize` it is a callback of commit() and  `CommitEntry` which contains Histories of committed `DictEntry` would be passed to it. `CommitEntry` would help to record commits to UserDictionary.

Reply:
Memory mostly did a lot of job handling UserDictionary for Userdictionary recording needs time tick which is handled by methods like `StartSession`  `FinishSession` or receiving deleted entry from FrontEnd or stops recording entry when facing unhandled key.